OpenRA.Mods.Common/Traits/BotModules/BotModuleLogic/BaseBuilderQueueManager.cs

```diff
@@ -134,14 +134,42 @@ bool TickQueue(IBot bot, ProductionQueue queue)
 				// HACK: HACK HACK HACK
 				// TODO: Derive this from BuildingCommonNames instead
 				var type = BuildingType.Building;
+				CPos? location = null;
+				string os = "PlaceBuilding";
 
-				// Check if Building is a defense and if we should place it towards the enemy or not.
-				if (world.Map.Rules.Actors[currentBuilding.Item].HasTraitInfo<AttackBaseInfo>() && world.LocalRandom.Next(100) < baseBuilder.Info.PlaceDefenseTowardsEnemyChance)
-					type = BuildingType.Defense;
-				else if (baseBuilder.Info.RefineryTypes.Contains(world.Map.Rules.Actors[currentBuilding.Item].Name))
-					type = BuildingType.Refinery;
+				///Check if Building is a plug for other Building
+				if (world.Map.Rules.Actors[currentBuilding.Item].HasTraitInfo<PlugInfo>())
+				{
+					var pluginfo = world.Map.Rules.Actors[currentBuilding.Item].TraitInfoOrDefault<PlugInfo>();
+					var possible_buildings = world.ActorsHavingTrait<Pluggable>().Where(
+						a => a.Owner == player &&
+						a.TraitsImplementing<Pluggable>().Where(
+							plugp => plugp.AcceptsPlug(a, pluginfo.Type)).Any());
+
+					var building = possible_buildings.FirstOrDefault();
+					var plugs = possible_buildings.FirstOrDefault().TraitsImplementing<Pluggable>().Where(port => port.AcceptsPlug(possible_buildings.FirstOrDefault(), pluginfo.Type));
+
+					if (plugs.Any())
+					{
+						os = "PlacePlug";
+						location = building.Location + plugs.FirstOrDefault().Info.Offset;
+					}
+					else
+					{
+						return false;
+					}
```

Comment:
This can simplify a lot by using the following:
```suggestion
					var possibleBuilding = world.ActorsWithTrait<Pluggable>().FirstOrDefault(a =>
						a.Actor.Owner == player && a.Trait.AcceptsPlug(a.Actor, plugInfo.Type));

					if (possibleBuilding != null)
					{
						os = "PlacePlug";
						location = possibleBuilding.Actor.Location + possibleBuilding.Trait.Info.Offset;
					}
					else
						return false;
```